src/backend/resources.rs

```diff
@@ -0,0 +1,228 @@
+//! estimates the total size of the generated hardware.
+
+use std::io::Write;
+use std::vec;
+use std::{collections::HashMap, io};
+
+use crate::backend::traits::Backend;
+use calyx::ir::Component;
+use calyx::{
+    errors::CalyxResult,
+    ir::{self, Binding, CellType, Id},
+    utils::OutputFile,
+};
+
+#[derive(Default)]
+pub struct ResourcesBackend;
+
+impl Backend for ResourcesBackend {
+    fn name(&self) -> &'static str {
+        "resources"
+    }
+
+    /// OK to run this analysis on any Calyx program
+    fn validate(_ctx: &ir::Context) -> CalyxResult<()> {
+        Ok(())
+    }
+
+    /// Don't need to take care of this for this pass
+    fn link_externs(
+        _ctx: &ir::Context,
+        _file: &mut OutputFile,
+    ) -> CalyxResult<()> {
+        Ok(())
+    }
+
+    fn emit(ctx: &ir::Context, file: &mut OutputFile) -> CalyxResult<()> {
+        let main_comp = ctx
+            .components
+            .iter()
+            .find(|comp| comp.name == ctx.entrypoint)
+            .unwrap();
+
+        let count_map: &mut HashMap<(Id, Binding), u32> = &mut HashMap::new();
+
+        gen_count_map(ctx, main_comp, count_map);
+
+        write_csv(count_map.clone(), file);
+
+        estimated_size(count_map.clone());
+
+        Ok(())
+    }
+}
+
+/// Counts the number of each primitive with a given set of parameters
+/// in the program with entrypoint `main_comp`.
+/// TODO: handle multi-component designs
+fn gen_count_map(
+    ctx: &ir::Context,
+    main_comp: &Component,
+    count_map: &mut HashMap<(Id, Binding), u32>,
+) {
+    for cell in main_comp.cells.iter() {
+        let cell_ref = cell.borrow();
+        let primitives = vec![
+            "std_reg",
+            "std_mem_d1",
+            "std_mem_d2",
+            "std_mem_d3",
+            "std_mem_d4",
+            "seq_mem_d1",
+            "seq_mem_d2",
+            "seq_mem_d3",
+            "seq_mem_d4",
+        ];
+        match &cell_ref.prototype {
+            CellType::Primitive {
+                name,
+                param_binding,
+                ..
+            } => {
+                if primitives.contains(&name.id.as_str()) {
+                    *count_map
+                        .entry((*name, (**param_binding).clone()))
+                        .or_insert(0) += 1;
+                }
+            }
+            CellType::Component { name } => {
+                let component = ctx
+                    .components
+                    .iter()
+                    .find(|comp| comp.name == name)
+                    .unwrap();
+                gen_count_map(ctx, component, count_map);
+            }
+            _ => (),
+        }
+    }
+}
+
+/// Writes a CSV to stdout with primitive count information
+/// generated by `gen_count_map`.
+/// TODO (priya): fix the output format and have it output to an actual file!
+fn write_csv(count_map: HashMap<(Id, Binding), u32>, _file: &mut OutputFile) {
```

Comment:
Just to be clear, I think it would be easiest (and most convenient) to write to `_file` here (rather than opening a new file or whatever). That will make this output go to wherever the compiler is supposed to write its output, including stdout. It should work to do `_file.get_write()` and use that in place of `io::stdout()` in the `csv::Writer::from_writer` call below.